src/Effect/Aff.js

```diff
@@ -475,8 +475,12 @@ var Aff = function () {
               // bracket count is non-zero then we are in a masked state so it's
               // impossible to be killed.
               if (interrupt && interrupt !== tmp && bracketCount === 0) {
-                step = attempt._1.killed(util.fromLeft(interrupt))(attempt._2);
-              } else if (fail) {
+               if (step == null || util.isLeft(step)) {
+                 step = attempt._1.killed(util.fromLeft(interrupt))(attempt._2);
+               } else {
+                 step = attempt._1.completed(util.fromRight(step))(attempt._2);
+               }
```

Comment:
this approach *almost* works, but falls over if the bracket is cancelled before the actual body run. In that case `step` will be some `Right` value returned from the bracket acquisition phase.

The other thing that is a bit annoying is that this loses the information that the bracket has been killed. It's almost like it needs to be `{ ..., killed :: Error -> Maybe a -> Aff Unit }`, but one thing at a time.

Reply:
The latest commit fixes the former, so I'll resolve this and mention the second problem as a comment on the PR again.

Reply:
There's still an issue with nested brackets. `step` may be mutated in the inner bracket since it's already masked. I am throwing the towel for the day.